### PR TITLE
Add support for X25519 keys and XEdDSA signature algorithm.

### PR DESCRIPTION
**Pull request by maraino**
### Description

This PR adds support for X25519 keys and the XEdDSA signature algorithm. It also adds support for parsing nebula public and private keys.

The X25519 keys are used by Nebula, and the XEdDSA algorithm is in https://signal.org/docs/specifications/xeddsa/

@wadey

### CODE REVIEW

File: x25519/x25519_test.go [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+			0x11, 0xc7, 0xf3, 0xe6, 0xc4, 0xdf, 0x9e, 0x8a,
+			0x51, 0x50, 0xe1, 0xdb, 0x3b, 0x30, 0xf9, 0x2d,
+			0xe3, 0xa3, 0xb3, 0xaa, 0x43, 0x86, 0x56, 0x54,
+			0x5f, 0xa7, 0x39, 0x0f, 0x4b, 0xcc, 0x7b, 0xb2,
+			0x6c, 0x43, 0x1d, 0x9e, 0x90, 0x64, 0x3e, 0x4f,
+			0x0e, 0xaa, 0x0e, 0x9c, 0x55, 0x77, 0x66, 0xfa,
+			0x69, 0xad, 0xa5, 0x76, 0xd6, 0x3d, 0xca, 0xf2,
+			0xac, 0x32, 0x6c, 0x11, 0xd0, 0xb9, 0x77, 0x02,
+		}, false},
+		{"ok", []byte{
```
Comment (hslatman) [inline x25519/x25519_test.go:131]: Some additional info about the test case in the name may be useful.

File: x25519/x25519.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+	// The following code should be equivalent to:
+	//
+	//   pub, err := publicKey.ToEd25519()
+	//   if err != nil {
+	//       return false
+	//   }
+	//   return ed25519.Verify(pub, message, sig)
+
+	if l := len(publicKey); l != PublicKeySize {
+		panic("x25519: bad public key length: " + strconv.Itoa(l))
```
[the remark this answers is not in the document]
Comment (maraino) [reply x25519/x25519.go:208]: I'm imitating the standard library behavior
https://github.com/golang/go/blob/b357b05b70d2b8c4988ac2a27f2af176e7a09e1b/src/crypto/ed25519/ed25519.go#L150

File: x25519/x25519.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+//   xeddsa_sign(k, M, Z):
+//       A, a = calculate_key_pair(k)
+//       r = hash1(a || M || Z) (mod q)
+//       R = rB
+//       h = hash(R || A || M) (mod q)
+//       s = r + ha (mod q)
+//       return R || s
+func Sign(rand io.Reader, p PrivateKey, message []byte) (signature []byte, err error) {
+	if l := len(p); l != PrivateKeySize {
+		panic("x25519: bad private key length: " + strconv.Itoa(l))
```
[the remark this answers is not in the document]
Comment (maraino) [reply x25519/x25519.go:122]: I'm imitating the standard library behavior
https://github.com/golang/go/blob/b357b05b70d2b8c4988ac2a27f2af176e7a09e1b/src/crypto/ed25519/ed25519.go#L150

File: x25519/x25519.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+//
+//   xeddsa_sign(k, M, Z):
+//       A, a = calculate_key_pair(k)
+//       r = hash1(a || M || Z) (mod q)
+//       R = rB
+//       h = hash(R || A || M) (mod q)
+//       s = r + ha (mod q)
+//       return R || s
+func Sign(rand io.Reader, p PrivateKey, message []byte) (signature []byte, err error) {
+	if l := len(p); l != PrivateKeySize {
```
[the remark this answers is not in the document]
Comment (maraino) [reply x25519/x25519.go:121]: l is reused in the panic

File: x25519/x25519.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+func Verify(publicKey PublicKey, message []byte, sig []byte) bool {
+	// The following code should be equivalent to:
+	//
+	//   pub, err := publicKey.ToEd25519()
+	//   if err != nil {
+	//       return false
+	//   }
+	//   return ed25519.Verify(pub, message, sig)
+
+	if l := len(publicKey); l != PublicKeySize {
```
[the remark this answers is not in the document]
Comment (maraino) [reply x25519/x25519.go:207]: l is reused in the panic

File: x25519/x25519_test.go (hunk 1 above)
Comment (maraino) [reply x25519/x25519_test.go:135]: I added references to the test vectors as comments. All those have to be ok.

File: x25519/x25519.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hslatman) [reply x25519/x25519.go:121]: My bad; sorry.

File: x25519/x25519.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hslatman) [reply x25519/x25519.go:207]: My bad; sorry.

File: x25519/x25519.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hslatman) [reply x25519/x25519.go:122]: Ah, I wasn't aware of that. I guess it's okay to panic then.

File: x25519/x25519.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hslatman) [reply x25519/x25519.go:208]: Ditto

File: x25519/x25519_test.go (hunk 1 above)
Comment (hslatman) [reply x25519/x25519_test.go:135]: Great 👍

**Review by hslatman — APPROVED** [review]

**Pull request merged**